### PR TITLE
Add support for sourced metrics.

### PR DESCRIPTION
**Pull request by romanb**
This is a little follow-up to the suggestion in https://github.com/paritytech/substrate/pull/6870 for [custom metric collectors](https://github.com/paritytech/substrate/pull/6870/files#r469152168). This PR introduces what I dubbed "sourced" counters or gauges via a custom collector.  A sourced counter or gauge is a metric that obtains its values from an existing source, i.e. not from a `prometheus`-crate metric. It thus allows feeding existing counters and gauges into prometheus without having to duplicate them as `prometheus`-crate counters or gauges and without having to declare counters as gauges or making extra gymnastics with the counter API (i.e. get() + compute diff + `inc_by()` or `reset()` + `inc_by()` for every update). The first use-case are the bandwidth counters from libp2p. I hope I understood and incorporated the labeling mechanisms from the `prometheus`-crate correctly. This is how the libp2p bandwidth counters show up in the localhost `/metrics` endpoint:
```
# TYPE substrate_sub_libp2p_network_bytes_total counter
substrate_sub_libp2p_network_bytes_total{direction="in"} 1808817
substrate_sub_libp2p_network_bytes_total{direction="out"} 36561
```

### CODE REVIEW

File: client/network/src/service.rs [hunk 1]
```diff
@@ -1164,7 +1167,34 @@ struct Metrics {
...
+	type N = u64;
+
+	fn collect(&self, mut set: impl FnMut(&[&str], Self::N)) {
+		set(&[&"in"], self.0.total_inbound());
+		set(&[&"out"], self.0.total_outbound());
+	}
+}
+
 impl Metrics {
+	fn register_bandwidth(registry: &Registry, sinks: Arc<transport::BandwidthSinks>)
```
Comment (mxinden) [inline client/network/src/service.rs:1184]: I am a bit torn whether this should be part of `Metrics` as in the end `Metrics` doesn't even keep a reference to the `BandwidthSource`. What do you think of doing the registration in the `// Initialize the metrics.` block above right away? Again, not sure that is easier to read.

File: utils/prometheus/src/sourced.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+					let mut g = proto::Gauge::default();
+					g.set_value(value.into_f64());
+					m.set_gauge(g);
+				}
+				t => {
+					log::error!("Unsupported sourced metric type: {:?}", t);
+				}
+			}
+
+			match self.desc.variable_labels.len().cmp(&label_values.len()) {
```
Comment (mxinden) [inline utils/prometheus/src/sourced.rs:89]: All rust-prometheus methods panic when providing unequal label keys and values. While not the greates UX, I would do the same (at least in debug mode) to stay consistent.

E.g. https://docs.rs/prometheus/0.7.0/prometheus/core/struct.MetricVec.html#method.with_label_values.

File: utils/prometheus/src/sourced.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+use prometheus::proto;
+use std::{cmp::Ordering, marker::PhantomData};
+
+/// A counter whose values are obtained from an existing source.
+pub type SourcedCounter<S> = SourcedMetric<Counter, S>;
+
+/// A gauge whose values are obtained from an existing source.
+pub type SourcedGauge<S> = SourcedMetric<Gauge, S>;
+
+/// The type of a sourced counter.
```
Comment (mxinden) [inline utils/prometheus/src/sourced.rs:29]: To make sure people won't shoot themselves in the foot, what do you think of something along the lines of:

```suggestion
/// The type of a sourced counter.
///
/// A Prometheus counter is a monotonically increasing counter whose value can only ever increase. This property is not enforced here and instead has to be enforced by the user of this crate.
```

File: client/network/src/service.rs (hunk 1 above)
Comment (romanb) [reply client/network/src/service.rs]: I still think it preferable to not bury the definition of a metric in other control flow. Is https://github.com/paritytech/substrate/pull/6895/commits/22be2f76fd5d07ae390b9d5ab92e86f0c8e251e2 acceptable to you? (I.e. `Metrics::register_bandwidth` -> `BandwidthCounters::register`).

File: utils/prometheus/src/sourced.rs (hunk 2 above)
Comment (romanb) [reply utils/prometheus/src/sourced.rs:94]: Done in https://github.com/paritytech/substrate/pull/6895/commits/45b734cce4708839bf290c0d872b4cb1d690fd52.

File: utils/prometheus/src/sourced.rs (hunk 3 above)
Comment (romanb) [reply utils/prometheus/src/sourced.rs:33]: I added a comment in https://github.com/paritytech/substrate/pull/6895/commits/d341e21a3f60a163d0012357b9caf589712662e4.

**Comment by romanb** [comment]
@mxinden Thanks for the quick review. Let me know if there is anything else.

File: client/network/src/service.rs [hunk 4]
```diff
@@ -353,6 +348,17 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkWorker<B, H> {
 			(builder.build(), bandwidth)
 		};
 
+		// Initialize the metrics.
+		let metrics = match &params.metrics_registry {
+			Some(registry) => {
+				// Sourced metrics.
+				BandwidthCounters::register(registry, bandwidth.clone())?;
```
Comment (mxinden) [inline client/network/src/service.rs:355]: :+1:

File: client/network/src/service.rs [hunk 5]
```diff
@@ -1168,7 +1171,37 @@ struct Metrics {
...
+	type N = u64;
+
+	fn collect(&self, mut set: impl FnMut(&[&str], Self::N)) {
+		set(&[&"in"], self.0.total_inbound());
+		set(&[&"out"], self.0.total_outbound());
+	}
+}
+
 impl Metrics {
+
```
Comment (mxinden) [inline client/network/src/service.rs:1204]:
```suggestion
```

**Pull request merged**